test/test_profiler.py

```diff
@@ -167,9 +167,12 @@ def __iter__(self):
                 return iter(self.datapipe)
 
         dp2 = IDPDelegator(dp1)
-        it_dp2 = iter(dp2)
-        self.assertEqual(5, it_dp2.get_value(5))
-        self.assertEqual(list(range(10)), list(it_dp2))
+        it_dp2 = iter(dp2)  # This creates a 2nd iterator, hence invalidating `it_dp1` and `dp1`
+        self.assertEqual(5, it_dp2.get_value(5))  # type: ignore[attr-defined]
+        with self.assertRaisesRegex(RuntimeError, "A separate iterator has been created"):
+            next(it_dp2)
+        with self.assertRaisesRegex(RuntimeError, "A separate iterator has been created"):
+            self.assertEqual(list(range(10)), list(it_dp2))
```

Comment:
The necessary change to this test case here is worth having a closer look.

The basic issue was that within our constraint logic, we decided to increment the counter even when `iter` returns `self`, and this invalidates the source datapipe `dp1` such that calling `next` will trigger an exception